mcf-browser/src/back-end/parse-mcf.js

```diff
@@ -0,0 +1,216 @@
+  }
+  // get node from a local Id
+  if (propValue.startsWith('l:') || !propValue.includes(':')) {
+    const localID = propValue.replace('l:', '').trim();
+    return Node.getNode(localID, /* shouldCreate */ true, /* isDCID */ false);
+  }
+  // get node from a dcid
+  const dcid = getDCIDFromPropValue(propValue);
+  if (dcid) {
+    return Node.getNode(dcid, /* shouldCreate */ true,
+        /* isDCID */ true);
+  }
+
+  console.log('ERROR <getNodeFromPropValue> : ' + propValue);
+  return null;
+}
+
+/**
+ * Parses a string represetnaing a comma separated list of property values from
+ * a line of an mcf file. Converts the string into an array where each element
+ * is each element is either a Node or string representing the property value.
+ *
+ * @param {string} propValues A comma separated list of property values.
+ * @return {Array<Node|string>} The list of parsed values as an array.
+ */
+function
+parsePropValues(propValues) {
+  const values = [];
+
+  for (const propValue of propValues.split(',')) {
+    switch (propValue.split(':').length) {
+      case 2:
+        const targetNode = getNodeFromPropValue(propValue, true);
+        if (targetNode) values.push(targetNode);
+        break;
+
+      case 1: const value = propValue.trim();
+        if (value) values.push(value);
+        break;
+
+      default: console.log('ERROR <parsePropValues> too many colons: ' +
+                               propValues);
+    }
+  }
+  return values;
+}
+/** Class responsible for parsing an mcf file. */
+class ParseMCF {
+  /**
+   * Create a ParseMCF object which keeps tracks of the current source node of
+   * each triple in the mcf and the provenance, which is the mcf file name.
+   */
+  constructor() {
+    this.curNode = null;
+    this.prov = null;
+  }
+
+  /**
+   * Parses a single line of an mcf file. First determines if the line should be
+   * read, then finds the property label, then the property values and creates
+   * an Assertion object based on the triplle, given the calling object has a
+   * curNode property that acts as the source  and a prov property that gives
+   * the provenance for the triple.
+   *
+   * @param {string} line The line of mcf to be parsed.
+   */
+  parseLine(line) {
+    line = line.trim();
+
+    if (!shouldReadLine(line)) return;
+
+    const propLabel = getPropLabel(line);
+    if (!propLabel) return;
+
+    const propValues = line.replace(propLabel + ':', '').trim();
+    if (!propValues) return;
+
+    switch (propLabel) {
+      case 'Node':
+        this.curNode = getNodeFromPropValue(propValues, false);
+        Node.nodeLocalHash[this.curNode.localID] = this.curNode;
+        break;
+
+      case 'dcid':
+        if (this.curNode) {
+          this.curNode.setDCID(propValues.replace(/"/g, ''));
+        } else {
+          console.log('ERROR dcid prop declared before Node:' + line);
+        }
+        break;
+
+      default:
+        const values = parsePropValues(propValues);
```

Comment:
The current logic of getNodeFromPropValue() with optional bool, and parsePropValues() feels a bit convoluted to me.

What do think instead of first parsing out the values into either a string or pair of namespace+reference, and then handling Node vs. reference cases separately, instead of combining them in a single function. Something like below (in pseudo code)...

```
  // parsePropValues() returns a list having either a string or <namespace, reference> pair.
  parsedValues = parsePropValues(propValues)
  switch (propLabel) {
    case 'Node':
        // assert len(parsedValues) is 1
        // create new node
        // if value is a pair, ensure namespace = dcid and call setDCID(reference)
    case 'dcid':
        // assert len(parsedValues) is 1
        // assert the value is string
        // do setDCID
     default:
        for (const v of parsedValues) {
           if (v is ns+ref pair) {
             if (ns == LOCAL_NS) {
                // add local reference Assertion
             } else {
                // add DCID reference Assertion
             }
           } else {
              // add value AddAssertion
           }
        }
  }  
```

Reply:
sounds good. I am working on this now.